<a href="https://colab.research.google.com/github/i8dacake/Mat-421/blob/main/ModuleD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Linear Algebra**

### **1.1 Introduction**

In STEM, one comes across a system of linear relationships. These relationships can be learnt through a comprehensive understanding of Linear Algebra.

It is essential to understand some fundamentals before diving into linear systems.

A **set** is a collection of objects and is denoted by braces {}. An  **empty set** contains no objects and is denoted by empty braces {}. Given two sets, M and N, M ∩ N is the **intersection** of M and N where the set contains all elements that belong to both M and N. On the other hand, M U N is the **union** of M and N where the set contains all the elements of A and B. When a ∈ M, we say that a is an object contained in the set M.

There are several standard sets that are related to numbers: **naturals** ({1, 2, 3, ...}) , **wholes** ({0, 1, 2, 3, ...}), integers ({..., -2, -1, 0, 1, 2, ...}), **rationals** ({𝑝/𝑞 : 𝑝 ∈ ℤ, 𝑞 ∈ ℤ ∖ {0}}), **irrationals** (the set of real numbers that cannot be expressed as a fraction of integers), **reals** (rational U irrational numbers) and **complex** ({𝑎 + 𝑏𝑖 : 𝑎, 𝑏 ∈ ℝ, 𝑖 = sqrt(-1)}).

A **vector** in $ℝ^n$ is an n-tuple, or point. It can be written horizontally (**row vector**) or vertically (**column vector**). If the context of a vector is vague, it can be assumed that it is a column vector. The **zero vector** is a vector containing all zeroes. The **norm** of a vector is the value of its length. The **determinant** is a special number that can be calculated from a matrix. Vectors can be **dependent**(scalar multiples of each other or determinant is equal to 0) or **independent**(not scalar multiples of each other or determinant is not equal to 0).  

A m x n **matrix** is a rectangular table of numbers consisting of m rows and n columns. The **identity matrix** is a square matrix with ones on the diagonal and zeroes everywhere else. The **inverse** N of a square matrix M has the same dimensions as the original matrix, such that M * N = I. If a matrix has an inverse, it is said to be **non-singular**, else **singular**. One of the important points to note about matrices is that matrix multiplcation is **not** commutative.


***Create a row vector and column vector and show the shape of the vectors.***

In [5]:
import numpy as np

vector_row = np.array([[1, 2, 3, 4]])
vector_column = np.array([[1], [2], [3], [4]])

print(vector_row.shape)
print(vector_column.shape)

(1, 4)
(4, 1)


*We can observe that the dimensions (1, 4) of the row vector is 1 row and 4 columns while the dimensions (4, 1) of the column vector is 4 rows and 1 column.*

Many properties of sets, vectors and matrices can be verified using Python. The following example will be verifying that matrix multiplication is not commutative.

***Let the Python matrices M = [[1,7],[2,3],[5,0]] and N = [[2,6,3,1],[1,2,3,4]]. Compute the matrix product of M and N. Show that the product of M and N will produce an error.***

In [6]:
M = np.array([[1, 7], [2, 3], [5, 0]])

N = np.array([[2, 6, 3, 1], [1, 2, 3, 4]])

print(M)
print(N)
print(np.dot(M, N))
np.dot(N, M)

[[1 7]
 [2 3]
 [5 0]]
[[2 6 3 1]
 [1 2 3 4]]
[[ 9 20 24 29]
 [ 7 18 15 14]
 [10 30 15  5]]


ValueError: shapes (2,4) and (3,2) not aligned: 4 (dim 1) != 3 (dim 0)

***Use Python to find the determinant of the matrix 𝑀 = [[0,2,1,3],[3,2,8,1],[1,0,0,3],[0,3,2,1]]. Use the np.eye function to produce a 4 × 4 identity matrix, 𝐼. Multiply 𝑀 by 𝐼 to show that the result is 𝑀.***

In [7]:
from numpy.linalg import det

M = np.array([[0,2,1,3],
             [3,2,8,1],
             [1,0,0,3],
             [0,3,2,1]])
print('M:\n', M)

print('Determinant: %.1f'%det(M))
I = np.eye(4)
print('I:\n', I)
print('M*I:\n', np.dot(M, I))

M:
 [[0 2 1 3]
 [3 2 8 1]
 [1 0 0 3]
 [0 3 2 1]]
Determinant: -38.0
I:
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
M*I:
 [[0. 2. 1. 3.]
 [3. 2. 8. 1.]
 [1. 0. 0. 3.]
 [0. 3. 2. 1.]]


### **1.2 Elements of Linear Algebra**

A **linear combination** is a new vector that is constructed by multiplying each vector in a set of vectors by a constant and adding the results. A **span** is the set of all linear combinations of a set of vectors.  

A linear subspace is a result of linear combination wherein it is closed under addition and multiplication and must contain the zero vector. A **linear subspace** is a span of set of vectors.

A **column space** is the span of the columns of A.

A list of vectors are **linearly independent** if they cannot be written as a linear combination of the others. And, if a list of vectors are not linearly independent, then they are **linearly dependent**.

A **basis** is a list of vectors that are linearly independent.

If U is a linear subspace of V, then any basis of U has the same number of elements. This number is known as the dimension of U written as dim(U). This is known as the **Dimension Theorem**. The number of elements is called the dimension of the vector space. If we use this number in the context of a matrix, then the dimension of a column space of A is called the **column rank** of A.

A list of vectors {u1,...,um} is **orthonormal** if the ui’s are pairwise orthogonal and each has norm 1, that is for all i and all j  ̸ = i, ⟨ui,uj⟩ = 0, and ∥ui∥ = 1.

The **Pythagorean Theorem**, let u,v ∈ V be orthogonal. Then, $∥u+ v∥^2$ = $∥u∥^2$ + $∥v∥^2$.

The **Cauchy-Schwarz** lemma os that for any u, v ∈ V, |⟨u, v⟩| ≤ ∥u∥∥v∥.

Suppose U be a linear subspace of V with orthonormal basis m1,..., mn, and let v V. Then, ||v-Pu v|| = ||v-u|| for any u > U is valid. This is the **Best Approximation Theorem.**

Let A ∈ R^(d×d) be a square matrix. Then λ ∈ R is an **eigenvalue** of A if there exists a nonzero vector x  ̸ = 0 such that Ax = λ x. Here, the vector x is referred to as an **eigenvector**. If a matrix A is symmetric, then any two eigenvectors from different eigenspaces are orthogonal.

***Given the row vectors 𝑣 = [0,3,2], 𝑤 = [4,1,1], and 𝑢 = [0,−2,0], write the vector 𝑥 = [−8,−1,4] as a linear combination of 𝑣, 𝑤, and 𝑢.***

In [13]:
v = np.array([[0, 3, 2]])
w = np.array([[4, 1, 1]])
u = np.array([[0, -2, 0]])
x = 3*v-2*w+4*u
print(x)

[[-8 -1  4]]


***Given the vectors 𝑣 = [0,2,0] and 𝑤 = [3,0,0], use the Numpy function cross to compute the cross product of v and w.***

In [14]:
v = np.array([[0, 2, 0]])
w = np.array([[3, 0, 0]])
print(np.cross(v, w))

[[ 0  0 -6]]


***Calculate the eigenvalues and eigenvectors for matrix 𝐴 = [0, 2 ; 2, 3].***

In [8]:
import numpy as np
from numpy.linalg import eig

a = np.array([[0, 2],
              [2, 3]])
w,v=eig(a)
print('E-value:', w)
print('E-vector', v)

E-value: [-1.  4.]
E-vector [[-0.89442719 -0.4472136 ]
 [ 0.4472136  -0.89442719]]


***Compute the eigenvalues and eigenvectors for matrix 𝐴 = [1, 2, 3; 4, 5, 6; 7, 8, 9].***

In [10]:
a = np.array([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]])
w,v=eig(a)
print('E-value:', w)
print('E-vector', v)

E-value: [ 1.61168440e+01 -1.11684397e+00 -1.30367773e-15]
E-vector [[-0.23197069 -0.78583024  0.40824829]
 [-0.52532209 -0.08675134 -0.81649658]
 [-0.8186735   0.61232756  0.40824829]]


### **1.3 Linear Regression**

Linear regression is quite frequently used in the industry to build models. The models depend linearly on their unknown parameters and therefore, are easier to fit as compared to models that are non-linearly related to their parameters.

**QR Decomposition** is a procedure used to solve the linear least squares problem where a matrix is broken down into two components, Q and R. The matrix A, which is to be decomposed, is transformed into the product of an orthogonal matrix Q and an upper triangular matrix R, A = Q * R. THe orthogonal matrix Q consists of columns that are orthogonal unit vectors while the upper triangular matrix R has all entries below the main diagonal as 0. The first step is to use the Gram-Schmidt algorithm to get an orthionormal basis span from a linearly independent set of span.

The sytem Ax = b is often inconsistent. We want to try to use Ax to approximate b. So, we need an overdetermined system where n > m. But this matrix is not invertible. Therefore, the **Least-Squares** method works around this by looking for an approximate solution that reduces the distance between the projected data points and the actual data points.